AlfrescoDocumentPickerFileProvider/Enumerators/AFPFolderEnumerator.m

```diff
@@ -0,0 +1,166 @@
+/*******************************************************************************
+ * Copyright (C) 2005-2018 Alfresco Software Limited.
+ *
+ * This file is part of the Alfresco Mobile iOS App.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ ******************************************************************************/
+
+#import "AFPFolderEnumerator.h"
+#import "AFPServerEnumerator+Internals.h"
+#import "CustomFolderService.h"
+
+@interface AFPFolderEnumerator()
+
+@property (nonatomic, strong) CustomFolderService *customFolderService;
+@property (nonatomic, strong) AlfrescoDocumentFolderService *documentService;
+
+@end
+
+@implementation AFPFolderEnumerator
+
+- (void)enumerateItemsForObserver:(id<NSFileProviderEnumerationObserver>)observer startingAtPage:(NSFileProviderPage)page
+{
+    self.observer = observer;
+    
+    [self setupSessionWithCompletionBlock:^(id<AlfrescoSession> session) {
+        self.customFolderService = [[CustomFolderService alloc] initWithSession:session];
```

Comment:
the use of self within completion blocks creates retain cycles which materialise in memory leaks. you can use the weakfy strategy to overcome this:
`__weak typeof(self) weakSelf = self;
 [self setupSessionWithCompletionBlock:^(id<AlfrescoSession> session) {
__strong typeof(self) strongSelf = weakSelf;

strongSelf.property = ....
[strongSelf method];
// work here
}];`